Pub sub example

maxglassie opened the pull request:
@kevinbader @mmacai 

## Description
This adds an architecture example to the /examples folder that uses RIG purely for pub / sub to connect to backend services. The backend service is basic Springboot with SpringKafka and uses EmbeddedKafka to test. 

The goal here is to show how to set up, in the essentials, this type of architecture using RIG. That is possible to completely decouple the backend and the frontend and get rid of REST calls to backend services. Part of that objective is to have an example Springboot app that has unit tests and can be used as a starting point, like, an app template. 

The README in the Java app explains how to get the thing running and demonstrates the roundtrip connection. Admittedly, there is no UI, the demo is in the CLI. 

## What to look out for
Please use the Readme in the Java app to run the architecture and see if it works for you. Also, note that the documentation I added to the intro maybe should live somewhere else and most likely should be better. I appreciate suggestions. Consider this a first draft.

Technical notes: we had versioning issues when we set up this architecture and so the RIG version in the docker file is hardcoded to 2.1.1. Not sure how to update this, since it looks like RIG has been updated since then, but happy to do so with direction. @mmacai may remember when we dealt with that.

- check that the code works on your machine.
- suggest implementation-code design/structure/readability improvements.
- suggest test-code design/structure/readability improvements.
- suggest documentation improvements.
- suggest language/writing improvements.

mmacai commented on examples/rig-kafka-pub-sub-example/example/README.md:
````diff
@@ -0,0 +1,309 @@
+
+Run the app 
+```bash
+java -jar target/example-0.0.1-SNAPSHOT.jar
````
Any reason to not use Docker image?

maxglassie replied:
No, other than laziness and lack of fluency in Docker. I am working on this and have added it to the Dockerfile, but I can't get the Springboot java image to interact with Kafka. Any ideas?

maxglassie replied:
Docker Compose file

maxglassie replied:
update - I got it to work

mmacai commented on examples/rig-kafka-pub-sub-example/docker-starter/docker-compose.yml:
```diff
@@ -0,0 +1,48 @@
+
+  reactive-interaction-gateway:
+    container_name: reactive-interaction-gateway
+    image: accenture/reactive-interaction-gateway:2.1.1
```
Let's remove tag `2.1.1` and use latest version.

kevinbader replied:
just a thought, but wouldn't it make sense to build the image from `../../../..`? This way it would be consistent with the rest of the code in the repository, regardless of what's currently published on Docker Hub..

maxglassie replied:
I will make the update according to @mmacai 's comment. I like Kevin's idea, but I'm not a Docker wiz and would need a little support to make sure I'm implementing doing that correctly.

maxglassie replied:
I ran into this issue once I updated to latest. Any ideas?

```
~/projects/reactive-interaction-gateway/examples/rig-kafka-pub-sub-example/docker-starter (pub-sub-example)⚡ ./start.sh
Removing network docker-starter_default
Creating network "docker-starter_default" with the default driver
Pulling example (example:latest)...
ERROR: The image for the service you're trying to recreate has been removed. If you continue, volume data could be lost. Consider backing up your data before continuing.

Continue with the new image? [yN]y
Pulling example (example:latest)...
ERROR: pull access denied for example, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
```

mmacai replied:
For example like this https://github.com/Accenture/reactive-interaction-gateway/blob/master/examples/tests/run_all_tests.sh#L14, switch to root, build image, switch back, run compose

maxglassie replied:
I did not implement this. It currently just pulls the image from Docker Hub.

mmacai replied:
Would you mind if I take a look and potentially commit the change?

mmacai commented on examples/rig-kafka-pub-sub-example/docker-starter/docker-compose.yml:
```diff
@@ -0,0 +1,48 @@
+      - KAFKA_BROKERS=kafka:9092
+      - API_HTTP_PORT=7010
+      - INBOUND_PORT=7000
+      - PROXY_CONFIG_FILE=proxy/rig-proxy.json
```
Can be just `/rig-proxy.json`.

mmacai commented on examples/rig-kafka-pub-sub-example/docker-starter/docker-compose.yml:
```diff
@@ -0,0 +1,48 @@
+      - 7000:7000
+      - 7010:7010
+    volumes:
+      - ./rig-proxy.json:/opt/sites/rig/lib/rig_inbound_gateway-2.1.1/priv/proxy/rig-proxy.json
```
Current version can do better lookup, so this should be enough `- ./rig-proxy.json:/rig-proxy.json`.

mmacai commented on examples/rig-kafka-pub-sub-example/docker-starter/docker-compose.yml:
```diff
@@ -0,0 +1,48 @@
+      - LOG_LEVEL=debug
+      - KAFKA_SOURCE_TOPICS=rig,rig-consumed
+      - PROXY_KAFKA_REQUEST_TOPIC=rig
+      # - PROXY_KAFKA_RESPONSE_TOPICS=rig
```
Let's remove commented code.

maxglassie replied:
done

mmacai commented on examples/rig-kafka-pub-sub-example/example/README.md:
````diff
@@ -0,0 +1,309 @@
+```yml
+  reactive-interaction-gateway:
+    container_name: reactive-interaction-gateway
+    image: accenture/reactive-interaction-gateway:2.1.1
````
Remove `:2.1.1`

mmacai commented on examples/rig-kafka-pub-sub-example/example/README.md:
```diff
@@ -0,0 +1,309 @@
+      - KAFKA_BROKERS=kafka:9092
+      - API_HTTP_PORT=7010
+      - INBOUND_PORT=7000
+      - PROXY_CONFIG_FILE=proxy/rig-proxy.json
```
`/rig-proxy.json`

mmacai commented on examples/rig-kafka-pub-sub-example/example/README.md:
````diff
@@ -0,0 +1,309 @@
+
+Body
+```JSON
+{
````
I would move this json to `rig-proxy.json`, so it's automatically loaded and user doesn't have to bother with registering the api.

maxglassie replied:
Done. Thanks. I left how to do it with an endpoint in the readme as reference.

mmacai commented on examples/rig-kafka-pub-sub-example/example/README.md:
````diff
@@ -0,0 +1,309 @@
+
+Body:
+```JSON
+{
````
Would be good to use the most recent format for this request (this one is still supported, but not recommended due to the different format compared to regular events):

```
{
    "id":"069711bf-3946-4661-984f-c667657b8d85",
    "type":"com.example",
    "time":"2018-04-05T17:31:00Z",
    "specversion":"0.2",
    "source":"/cli",
    "contenttype":"application/json",
    "rig": {
    "target_partition": "the-partition-key"
  },
    "data": { 
              "payload": { "payload": "payload", "number": 2 }
            }
  }
```

maxglassie replied:
updated. thanks!

kevinbader commented on docs/intro.md:
```diff
@@ -159,7 +159,19 @@ Routes are defined using a configuration file or by POSTing directly to the appl
-## Next Layer of Complexity: A Reference Architecture
-In the [/examples](https://github.com/Accenture/reactive-interaction-gateway/tree/master/examples) folder, there is an example architecture made with a React frontend, RIG, a Node backend, and a Kafka instance. Here's a chart demonstrating that reference architecture in abstract. Go to [/examples](https://github.com/Accenture/reactive-interaction-gateway/tree/master/examples) for more depth:
+## Another Architectural Option: Using RIG for Publish / Subscribe Communication with the Backend
+It is possible to send an HTTP request to RIG and have it publish and subscribe to Kafka, decoupling the front end from the backend entirely.
```
I once had a discussion around that with one of our technical writers that review our official blog posts. The proper way to do it "frontend" and "backend", but "front-end service" and "back-end service". So if it's used as a stand-alone noun, it is one word.

maxglassie replied:
Good clarification. And very interesting, I wonder why the hyphenation in one case and not the other? Something about being an adjective rather than a noun? I will update accordingly.

kevinbader replied:
I just looked up her explanation:

> I see back end, backend and back-end used interchangeably in all sorts of contexts, technical or not. You have four instances of backend and four of frontend in your post. The words should be hyphenated if used as a compound adjective in front of another word. For example, you should have back-end services and back-end team but not HTTP backends. Same for front-end team and front-end code but not frontends or users’ frontends.

maxglassie commented on docs/intro.md:
```diff
@@ -159,7 +159,19 @@ Routes are defined using a configuration file or by POSTing directly to the appl
+
+Another shows the above publish and subscribe architecture, called `rig-kafka-pub-sub-example`. It uses Springboot for the backend microservice. Here's a diagram demonstrating that example in abstract:
+
+![examples-rig-kafka-pub-sub](assets/examples-rig-kafka-pub-sub.png)
```
lol yea, that is annoying. good catch.

maxglassie commented on examples/rig-kafka-pub-sub-example/example/.mvn/wrapper/MavenWrapperDownloader.java:
```diff
@@ -0,0 +1,114 @@
+/*
```
@kevinbader @krekac does this file feel like it should be in the gitignore? I don't know mvn well enough...

maxglassie replied:
I updated the RIG .gitignore for these things, hoping it works properly.

arana3 replied:
Gradle is considered spiritual successor to Maven. And with Gradle it is recommended for projects to use wrapper to ensure consistent version across developers and machines. It is part of source control (the wrapper files). Therefore, for Maven, I personally think it would hold true.

Though since you put the Maven package step inside the Dockerfile, executed via a base image that already contains the appropriate JDK and Maven version, I'd not be too worried about above.

maxglassie commented:
I updated my PR to resolve most of the comments (thanks for the feedback :)
). 

A few notes - I updated the RIG gitignore to try to stop .mvn / build binaries from java and idea from causing problems, but that may not be the right approach. Secondly, I had trouble building the RIG docker image from source code rather than pulling. It kept downloading the image even after I did the setup. It is likely a simple issue in my Docker Compose but I'm not sure. 

Right now, it works as is and even, I asked someone else to test on their machine too!

mmacai commented on examples/rig-kafka-pub-sub-example/example/README.md:
````diff
@@ -0,0 +1,260 @@
+
+Body:
+```JSON
+{
````
Can you please update the formatting here? Looks a bit scuffed and hard to read.

mmacai commented:
Verified again and seems to work pretty well. If you don't mind I would take a look at the issue with image you had. Other than that just one comment from my side and merge conflict in `.gitignore`.

arana3 commented:
I took a look at this PR. I liked the idea of having a Java based example. Ran it locally following the provided instructions. Some notes:

- Tests do not pass consistently. I got a 'Session Timeout' exception when running one of the tests that leverage the embedded Kafka server. These tests are more like integration tests (seem solid though). Need to ensure there are enough safeguards to make sure tests pass consistently regardless of unavoidable hiccups in test harness.
- _sync-publish_ API returns 504 status code. No particular error in RIG logs though

krekac commented on examples/rig-kafka-pub-sub-example/example/src/main/java/com/rig/example/model/EventPayload.java:
```diff
@@ -0,0 +1,16 @@
+    public String payload;
+    public Integer number;
+
+    public EventPayload transform() {
```
I would probably rather do this transform in the consumePolicyOwner method, as it is more like "business related operation" and there is no need to have it defined on the model

maxglassie commented on examples/rig-kafka-pub-sub-example/example/src/test/resources/application.yml:
```diff
@@ -0,0 +1,18 @@
+
+spring:
```
updated. but I removed the tests now so arguably, no longer necessary at all. thanks.

maxglassie replied:
Updated. Thanks!

maxglassie commented on examples/rig-kafka-pub-sub-example/example/src/main/java/com/rig/example/engine/Producer.java:
```diff
@@ -0,0 +1,20 @@
+@Service
+public class Producer {
+
+    private static final String TOPIC = "rig-consumed";
```
You're correct. I'm going to leave it hard-coded for now. We can update it in the future.

maxglassie commented on examples/rig-kafka-pub-sub-example/example/src/test/java/com/rig/example/engine/ConsumerTest.java:
```diff
@@ -0,0 +1,121 @@
+@RunWith(SpringRunner.class)
+@SpringBootTest
+@DirtiesContext
+public class ConsumerTest {
```
I have deleted the tests based on our conversations around licensing.
